0004-ecs2ext-protocol/README.md

```diff
@@ -0,0 +1,251 @@
+---
+name: Communication protocol between Endpoint Communication Service and extension service
+shortname: 5/ECS2EXT
+status: raw
+editor: Dmitry Sergeev <dsergeev@cybervisiontech.com>
+---
+
+## Introduction
+
+The ECS2EXT protocol is a [Kaa event-based protocol](/) extension.
+
+It is intended to solve the problem of communication between ECS and any extension service.
+
+## Requirements and constraints
+### Problems and possible solutions
+
+1. _Message delivery confirmation._
+
+   Solutions:
+   - Protocol contains field named "status" that is used to send error message or confirm successful delivery. 
+   
+2. _Ability to determine which Avro schema should be used on extension to deserialize `payload` field._ 
+
+   Solutions:
+   - MQTT path (e.g. "config/pull", "log/send" etc) will be passed as a `path` field in ECS2EXT protocol.
+
+3. _Ability to send payload in two directions._ 
+
+   Solutions:
+   - Protocol message contains `payload` field for both directions: from ECS to extension and vice versa.
+   
+4. _Ability to send acknowledgement (status) message without payload._
+
+   Solutions:
+   - `payload` and `contentType` fields are optional.
+   
+## Use cases
+
+### UC1
+Endpoint can send some data to extension that should not respond with any data excepting status message for delivery confirmation.
+In that case `payload` and `contentType` fields will be filled only in messages that goes from ECS to extension. Message from extension to ECS should contain `status` field filled, `payload` and `contentType` can be dropped.
+Example of extension with such communication:```Data collection extension```
+
+### UC2
+Endpoint can receive messages from extension with `payload` but respond with `status` message only.
+Example of such case is push notifications from server to client.
+
+### UC3
+Endpoint and extension can send and receive messages with `payload`.
+
+
+### Formats
+#### Avro JSON
+![](ecs2ext-ipc.png?raw=true)
+##### Message from ECS to extension
+ECS should send message using NATS to the next subject:
+```
+kaa.{protocol-version}.service.{service-instance-id}
+```
+Refer to [Kaa event-based protocol documentation](/) for `protocol-version` and `service-instance-id` parameters.
+
+Format of the ECS-to-extension message:
+- `correlationId` (string, required) - refer to [Kaa event-based protocol documentation](/) for description.
+- `timestamp` (number, required) - message creation timestamp.
+- `timeout` (number, required) - amount of milliseconds from `timestamp` 
+- `clientSessionId` (string, optional) - refer to [??](/) for description.
```

Comment:
I think we should remove this field from message. It is currently present in ECS to DCX message. But in fact extension has nothing to do with this data because it makes sense only on ECS side